ml-agents-envs/mlagents/envs/timers.py

```diff
@@ -0,0 +1,172 @@
+distinct blocks, and are tracked separately.
+"""
+
+
+class TimerNode:
+    """
+    Represents the time spent in a block of code.
+    """
+
+    __slots__ = ["children", "total", "count"]
+
+    def __init__(self):
+        self.children: Dict[str, "TimerNode"] = {}
+        self.total: float = 0.0
+        self.count: int = 0
+
+    def get_child(self, name: str) -> "TimerNode":
+        """
+        Get the child node corresponding to the name (and create if it doesn't already exist).
+        """
+        child = self.children.get(name)
+        if child is None:
+            child = TimerNode()
+            self.children[name] = child
+        return child
+
+    def add_time(self, elapsed: float) -> None:
+        """
+        Accumulate the time spent in the node (and increment the count).
+        """
+        self.total += elapsed
+        self.count += 1
+
+
+class TimerStack:
+    """
+    Tracks all the time spent. Users shouldn't use this directly, they should use the contextmanager below to make
+    sure that pushes and pops are already matched.
+    """
+
+    __slots__ = ["root", "stack"]
+
+    def __init__(self):
+        self.root = TimerNode()
+        self.stack = [self.root]
+
+    def push(self, name: str) -> TimerNode:
+        """
+        Called when entering a new block of code that is timed (e.g. with a contextmanager).
+        """
+        current_node: TimerNode = self.stack[-1]
+        next_node = current_node.get_child(name)
+        self.stack.append(next_node)
+        return next_node
+
+    def pop(self) -> None:
+        """
+        Called when exiting a new block of code that is timed (e.g. with a contextmanager).
+        """
+        self.stack.pop()
+
+    def get_timing_tree(self, node: TimerNode = None) -> Dict[str, Any]:
+        """
+        Recursively build a tree of timings, suitable for output/archiving.
+        """
+        if node is None:
+            node = self.root
+
+        res: Dict[str, Any] = {"total": node.total, "count": node.count}
+
+        child_total = 0.0
+        child_list = []
+        for child_name, child_node in node.children.items():
+            child_res: Dict[str, Any] = {
+                "name": child_name,
+                **self.get_timing_tree(child_node),
+            }
+            child_list.append(child_res)
+            child_total += child_res["total"]
+
+        # "self" time is total time minus all time spent on children
+        res["self"] = max(0.0, node.total - child_total)
+        if child_list:
+            res["children"] = child_list
+
+        return res
+
+
+# Global instance of a TimerStack. This is generally all that we need for profiling, but you can potentially
+# create multiple instances and pass them to the contextmanager
+_global_timer_stack = TimerStack()
+
+
+@contextmanager
+def hierarchical_timer(name: str, timer_stack: TimerStack = None) -> Generator:
```

Comment:
Should this just be called "timer" or is that too generic?

Reply:
I think this is good, better to be precise to help people understand.